sources/kafka/helpers.py

```diff
@@ -0,0 +1,215 @@
+from typing import Any, Dict, List
+
+from confluent_kafka import Consumer, Message, TopicPartition, OFFSET_BEGINNING  # type: ignore
+from confluent_kafka.admin import TopicMetadata  # type: ignore
+
+from dlt.common import pendulum
+from dlt.common.configuration import configspec
+from dlt.common.configuration.specs import CredentialsConfiguration
+from dlt.common.time import ensure_pendulum_datetime
+from dlt.common.typing import DictStrAny, TSecretValue, TAnyDateTime
+from dlt.common.utils import digest128
+
+
+def default_message_processor(msg: Message) -> Dict[str, Any]:
+    """Basic Kafka message processor.
+
+    Returns the message value and metadata. Timestamp consists of two values:
+    (type of the timestamp, timestamp). Type represents one of the Python
+    Kafka constants:
+        TIMESTAMP_NOT_AVAILABLE - Timestamps not supported by broker.
+        TIMESTAMP_CREATE_TIME - Message creation time (or source / producer time).
+        TIMESTAMP_LOG_APPEND_TIME - Broker receive time.
+
+    Args:
+        msg (confluent_kafka.Message): A single Kafka message.
+
+    Returns:
+        dict: Processed Kafka message.
+    """
+    ts = msg.timestamp()
+    topic = msg.topic()
+    partition = msg.partition()
+    key = msg.key().decode("utf-8")
+
+    return {
+        "_kafka": {
+            "partition": partition,
+            "topic": topic,
+            "key": key,
+            "offset": msg.offset(),
+            "ts": {
+                "type": ts[0],
+                "value": ensure_pendulum_datetime(ts[1] / 1e3),
+            },
+            "data": msg.value().decode("utf-8"),
+        },
+        "_kafka_msg_id": digest128(topic + str(partition) + str(key)),
```

Comment:
nothing against that. is `key()` really not globally unique though?

Reply:
It's not. Still, keys are used to calc hashes, which allows to decide into which partition the message must be saved.

![Untitled](https://github.com/dlt-hub/verified-sources/assets/26030071/0c6e5f1c-3248-42e3-8a36-3709aa62d8ab)